ddl/ddl_api.go

```diff
@@ -903,31 +903,37 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 		return errors.Trace(err)
 	}
 
-	if pi != nil && pi.Type == model.PartitionTypeRange {
-		// Only range type partition is now supported.
-		// Range columns partition only implements the parser, so it will not be checked.
-		if s.Partition.ColumnNames == nil {
-			if err = checkPartitionNameUnique(tbInfo, pi); err != nil {
-				return errors.Trace(err)
-			}
+	if pi != nil {
+		if pi.Type == model.PartitionTypeRange {
+			// Only range type partition is now supported.
+			// Range columns partition only implements the parser, so it will not be checked.
+			if s.Partition.ColumnNames == nil {
+				if err = checkPartitionNameUnique(tbInfo, pi); err != nil {
+					return errors.Trace(err)
+				}
 
-			if err = checkCreatePartitionValue(ctx, tbInfo, pi, cols); err != nil {
-				return errors.Trace(err)
-			}
+				if err = checkCreatePartitionValue(ctx, tbInfo, pi, cols); err != nil {
+					return errors.Trace(err)
+				}
 
-			if err = checkAddPartitionTooManyPartitions(len(pi.Definitions)); err != nil {
-				return errors.Trace(err)
-			}
+				if err = checkAddPartitionTooManyPartitions(uint64(len(pi.Definitions))); err != nil {
+					return errors.Trace(err)
+				}
 
-			if err = checkPartitionFuncValid(ctx, tbInfo, s.Partition.Expr); err != nil {
-				return errors.Trace(err)
-			}
+				if err = checkPartitionFuncValid(ctx, tbInfo, s.Partition.Expr); err != nil {
+					return errors.Trace(err)
+				}
 
-			if err = checkPartitionFuncType(ctx, s, cols, tbInfo); err != nil {
-				return errors.Trace(err)
-			}
+				if err = checkPartitionFuncType(ctx, s, cols, tbInfo); err != nil {
+					return errors.Trace(err)
+				}
 
-			if err = checkRangePartitioningKeysConstraints(ctx, s, tbInfo, newConstraints); err != nil {
+				if err = checkRangePartitioningKeysConstraints(ctx, s, tbInfo, newConstraints); err != nil {
+					return errors.Trace(err)
+				}
+			}
+		} else if pi.Type == model.PartitionTypeHash {
```

Comment:
Can I use `switch` here? There may be more partition types in the future.

Reply:
`switch` can't help for the indent.
maybe extract the `if pi.Type == model.PartitionTypeRange {` branch into a function makes the code look better.

Reply:
Support for extraction to another function.